TLS13 HelloRetryRequest Implementation

ryru opened the pull request:
Support for HelloRetryRequest (HRR) handling and enhancement in the test suite to verify connection with all supported curves.

Maybe we should discuss these two points:

- Server does not currently send a ChangeCipherSpec message after a HelloRetryRequest
- Curves can only be set via global settings currently, I do not see any way how to make test cases to automatically verify HRR-functionality.

Regarding the last bullet, the functionality could be implemented by adding an additional setting for curves to the `tls_socket_create`. Is this worthwile or do you see another way to add test suite capability? Would it also make sense to add cipher suite configuration support via `tls_socket_create`?

Thank you in advance for reviewing this change.

Commit:
tls-server: Handle HelloRetryRequest (HRR)

Adds support to handle retries with different DH group.

Commit:
tls-crypto: Add config file support for different DH groups to configure

and add new test cases to test TLS 1.3 connection for each supportd DH group.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -1537,7 +1567,8 @@ CALLBACK(group_filter, bool,
 		{
-			if (curves[i].group == group)
+			if (curves[i].group == group &&
+				filter_curve_config_suites(curves[i].curve))
```
I'd split this, i.e. do the filtering in `create_ec_enumerator()`. Because if you filter the groups here, configs will affect the unit tests (i.e. `tls_crypto_get_supported_curves()` will not return the _supported_ curves, but those supported _and configured_).

ryru replied:
I can do the refactoring of `filter_curve_config_suites()` to be more conform with the other config filters and work over an array. Although I currently do not see how I can do the filtering in `create_ec_enumerator()`. Let's discuss today.

ryru replied:
In my latest commits I changed the behaviour and add a note. see below.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -1091,6 +1091,36 @@ static void filter_specific_config_suites(private_tls_crypto_t *this,
+/**
+ * Filter key exchange curves by curve user config
+ */
+static bool filter_curve_config_suites(tls_named_group_t curve)
```
Why `_suites`?

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -1091,6 +1091,36 @@ static void filter_specific_config_suites(private_tls_crypto_t *this,
+	char *token, *config;
+	bool success;
+
+	success = TRUE;
```
That's one of the cases where assignment with the declaration is fine. Although, I actually think you don't really need this variable (you could just return `TRUE` when you find a match and `!config` in the other case, or you maybe could refactor this so the filtering operates on an array of groups, similar to the suites).

ryru replied:
In my latest commits I changed the behaviour and add a note. see below.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -1091,6 +1091,36 @@ static void filter_specific_config_suites(private_tls_crypto_t *this,
+	bool success;
+
+	success = TRUE;
+	config = lib->settings->get_str(lib->settings, "%s.tls.curve", NULL, lib->ns);
```
I wonder if it would be better to change `curve` to `group` or `dh_group` (not only here, but generally). Because at least in theory this also includes the classic DH groups e.g. _modp2048_ (called _ffdhe2048_ in the TLS 1.3 RFC), which is why the corresponding extension was renamed to "supported groups". Do other TLS clients actually support those other groups? Or only ECDH?

ryru replied:
As discussed we currently stick to _curve_ and postpone a more general renaming. I created a new issue #22.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -1091,6 +1091,36 @@ static void filter_specific_config_suites(private_tls_crypto_t *this,
+		enumerator = enumerator_create_token(config, ",", " ");
+		while (enumerator->enumerate(enumerator, &token))
+		{
+			if (enum_from_name(tls_named_group_names, token, &c) && curve == c)
```
I wonder it if would make sense to use names that are more consistent with those for IKE (e.g. _modp2048_ instead of _ffdhe2048_ or _ecp256_ instead of _secp256r1_). Parsing the token via `lib->proposal->get_token()` could be an option. `type` of the returned `const proposal_token_t` would have to be `DIFFIE_HELLMAN_GROUP` to be considered and the `algorithm` identifier would have to be converted from `diffie_hellman_group_t` to `tls_named_group_t`, that is, `ec_group_to_curve()` in `tls_server.c` could be refactored into a helper function in `tls_crypto.c` (properly declared in `tls_crypto.h` in this case).

ryru replied:
I migrated `ec_group_to_curve()` to `tls_ec_group_to_curve()` in `tls_crypto.c`. We can discuss this today.

ryru replied:
In my latest commits I changed the behaviour and add a note. see below.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -2321,3 +2352,33 @@ int tls_crypto_get_supported_suites(bool null, tls_version_t version,
+	enumerator =  enumerator_create_filter(
+							lib->crypto->create_dh_enumerator(lib->crypto),
+							group_filter, NULL, NULL);
+	while(enumerator->enumerate(enumerator, NULL, &current))
```
```suggestion
	while (enumerator->enumerate(enumerator, NULL, &current))
```

tobiasbrunner commented on src/libtls/tls_crypto.h:
```diff
@@ -669,4 +669,12 @@ tls_crypto_t *tls_crypto_create(tls_t *tls, tls_cache_t *cache);
 									tls_cipher_suite_t **suites);
 
+/**
+ * Get a list of all supported TLS ECDH groups.
```
```suggestion
 * Get a list of all supported TLS DH groups.
```
See the other comment regarding curves vs. groups, would also affect other comments here and the naming of the function/argument.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -456,43 +478,66 @@ static status_t process_client_hello(private_tls_server_t *this,
+		{
+			if (retrying(this))
+			{
+				DBG1(DBG_TLS, "already replied a retry request to client hello");
```
```suggestion
				DBG1(DBG_TLS, "already replied with a retry request to client hello");
```

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -456,43 +478,66 @@ static status_t process_client_hello(private_tls_server_t *this,
 			if (curve == key_type && peer_supports_curve(this, curve))
 			{
+				DBG1(DBG_TLS, "key share agreed on %N",
+		 			 tls_named_group_names, curve);
```
```suggestion
					 tls_named_group_names, curve);
```

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -1438,6 +1504,12 @@ METHOD(tls_handshake_t, change_cipherspec, void,
 
+	if (retrying(this))
+	{	/* client might send a ChangeCipherSpec after a HelloRetryRequest and
+ 		 * before a new ClientHello which should not cause any state changes */
```
```suggestion
		 * before a new ClientHello which should not cause any state changes */
```

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -44,6 +44,7 @@ typedef enum {
 	STATE_CIPHERSPEC_CHANGED_OUT,
 	STATE_FINISHED_SENT,
 	/* new states in TLS 1.3 */
+	STATE_HRR_SENT,
```
Can't you just go back to `STATE_INIT`?

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -162,6 +163,11 @@ struct private_tls_server_t {
+	/**
+	 * Is a HelloRetryRequest needed?
+	 */
+	bool need_retry_request;
```
Couldn't `requested_curve` be used for this? If we have a DH instance, `this->dh` could be queried for the actually negotiated group (converted via the helper mentioned in the comments for the previous commit).

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -308,9 +322,17 @@ static status_t process_client_hello(private_tls_server_t *this,
+				if (this->hashsig.len)
+				{
+					free(this->hashsig.ptr);
+				}
```
```suggestion
				chunk_free(&this->hashsig);
```

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -308,9 +322,17 @@ static status_t process_client_hello(private_tls_server_t *this,
+				if (this->curves.len)
+				{
+					free(this->curves.ptr);
+				}
```
```suggestion
				chunk_free(&this->curves);
```

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -456,43 +478,66 @@ static status_t process_client_hello(private_tls_server_t *this,
+			}
 			if (curve == key_type && peer_supports_curve(this, curve))
 			{
+				DBG1(DBG_TLS, "key share agreed on %N",
```
This log message sounds a bit weird.

ryru replied:
```suggestion
				DBG1(DBG_TLS, "DH group %N negotiated",
```
what do you think about this?

ryru replied:
`using key exchange %N`

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -456,43 +478,66 @@ static status_t process_client_hello(private_tls_server_t *this,
+				curve != key_type &&
+				peer_supports_curve(this, curve))
+			{
+				this->requested_curve = curve;
```
Use a local variable here if `requested_curve` is used as indicator for a requested/sent retry.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -456,43 +478,66 @@ static status_t process_client_hello(private_tls_server_t *this,
 			}
 		}
 		enumerator->destroy(enumerator);
 		if (!this->dh)
```
I'd add an empty line before this if statement because there a lot more code in there now.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -456,43 +478,66 @@ static status_t process_client_hello(private_tls_server_t *this,
-			key_type != TLS_CURVE448)
-		{	/* classic format (see RFC 8446, section 4.2.8.2) */
-			if (key_share.ptr[0] != TLS_ANSI_UNCOMPRESSED)
+			DBG1(DBG_TLS, "no suitable key share found in client hello");
```
Note that currently only the first key share extension is actually processed.

ryru replied:
That's right, good point. I suggest to add a comment here to emphasise it. What do you think about `currently only the first key share extensions is processed other offered key shares are ignored.`?

ryru replied:
I add a TODO with the above comment in the code and created a new issue #21 to track this feature request.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -456,43 +478,66 @@ static status_t process_client_hello(private_tls_server_t *this,
 			chunk_clear(&shared_secret);
-			return NEED_MORE;
+			this->requested_curve = key_type;
+			this->need_retry_request = FALSE;
```
Not necessary if `requested_curve` is used to indicate whether a retry is needed.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -975,22 +1029,31 @@ static status_t send_server_hello(private_tls_server_t *this,
+		}
+		else
+		{
+			if (!tls_write_key_share(&key_share, this->requested_curve, this->dh))
```
As mentioned, we don't really need `requested_curve` here, the identifier can easily be determined via `this->dh` and a helper function.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -975,22 +1029,31 @@ static status_t send_server_hello(private_tls_server_t *this,
-			this->alert->add(this->alert, TLS_FATAL, TLS_INTERNAL_ERROR);
-			extensions->destroy(extensions);
-			return NEED_MORE;
+			DBG1(DBG_TLS, "requests key exchange with %N",
```
```suggestion
			DBG1(DBG_TLS, "requesting key exchange with %N",
```

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -1375,7 +1438,10 @@ METHOD(tls_handshake_t, build, status_t,
 			case STATE_HELLO_SENT:
 			case STATE_CIPHERSPEC_CHANGED_OUT:
-				return send_encrypted_extensions(this, type, writer);
+				if (!this->need_retry_request)
```
Why is this check necessary?

tobiasbrunner reviewed (changes requested):
>     * Server does not currently send a ChangeCipherSpec message after a HelloRetryRequest

For our most common use cases this won't be necessary as TLS will be encapsulated in EAP anyway.

>     * Curves can only be set via global settings currently, I do not see any way how to make test cases to automatically verify HRR-functionality.

Yeah, that's tricky. In particular because those lists are filtered at runtime, not when the sockets are created (as was the case earlier with suites - i.e. suite negotiation changed too).

> Regarding the last bullet, the functionality could be implemented by adding an additional setting for curves to the `tls_socket_create`. Is this worthwile or do you see another way to add test suite capability? Would it also make sense to add cipher suite configuration support via `tls_socket_create`?

Don't really like adding even more arguments to that constructor. Currently don't see an easy way to do this.

Commit:
libtls: Add TLS version negotiation test cases

These client to server negotiation test cases are implemented:
- TLS 1.0 => TLS 1.3
- TLS 1.1 => TLS 1.3
- TLS 1.2 => TLS 1.3
- TLS 1.3 => TLS 1.3

- TLS 1.3 => TLS 1.0
- TLS 1.3 => TLS 1.1
- TLS 1.3 => TLS 1.2
- TLS 1.3 => TLS 1.3

- TLS 1.0 => TLS 1.2
- TLS 1.1 => TLS 1.2
- TLS 1.2 => TLS 1.2
- TLS 1.3 => TLS 1.2

Commit:
tls-crypto: Generalising DH group to TLS group mapping

and simplifying interface of key share extension writing by using the new generalised group mapping function.

ryru commented on src/libtls/tests/suites/test_socket.c:
```diff
@@ -442,6 +442,63 @@ static void test_tls(tls_version_t version, uint16_t port, bool cauth, u_int i)
+
+	count = tls_crypto_get_supported_curves(&curves);
+	ck_assert(i < count);
+	snprintf(curve, sizeof(curve), "%N", dh_group_config_names, curves[i]);
```
Maybe there is a better more generic way to map `diffie_hellman_group_names` such as `ECP_256` to `ecp256` as `get_token()` expects it. I noted also that the names has to be in lower case.

tobiasbrunner replied:
Yeah, that's not ideal. We could perhaps add a `diffie_hellman_group_short_names` like we have for e.g. hash algorithms that contains the proposal keywords of these algorithms. Or we could add aliases for the `diffie_hellman_group_names` (i.e. with underscore, see below for the case problem) to the static keyword list.

> I noted also that the names has to be in lower case.

Interesting, didn't know that (although, our proposals and configs are generally lower case). I guess the `-c` option for `gperf` isn't enough, probably also requires `--ignore-case`.

ryru replied:
Yes `--ignore-case` is required.

ryru commented on src/libtls/tls_crypto.c:
```diff
@@ -1554,12 +1584,46 @@ CALLBACK(group_filter, bool,
+							enumerator_create_filter(
+								lib->crypto->create_dh_enumerator(lib->crypto),
+								group_filter, NULL, NULL),
+							config_filter, NULL, NULL);
```
I spilt the two filters as you mentioned, but technically there is no difference in what `create_ec_enumerator` returns, is it? The callback `group_filter` is only used from within this method.

tobiasbrunner replied:
That's incorrect, though, because `lib->crypto->create_dh_enumerator()` and `group_filter()`  should also be used in `tls_crypto_get_supported_curves()` so only curves/groups are returned that are actually supported by the loaded plugins (otherwise the tests will fail if e.g. no plugin provides _curve448_).

tobiasbrunner commented on src/libtls/tls_peer.c:
```diff
@@ -1361,7 +1359,7 @@ static status_t send_client_hello(private_tls_peer_t *this,
 		DBG2(DBG_TLS, "sending extension: %N",
```
Maybe the `this->dh` check right before this should be removed, so that the initiation fails via `tls_write_key_share()` if the DH group is not supported (or does sending the client hello without key share make sense?).

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -911,20 +911,23 @@ METHOD(tls_handshake_t, process, status_t,
 
-	if (!dh || !dh->get_my_public_value(dh, &pub))
+	curve = tls_ec_group_to_curve(dh->get_dh_group(dh));
+	if (!dh ||
```
This check has to be done before getting the group from the object.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -1091,6 +1091,36 @@ static void filter_specific_config_suites(private_tls_crypto_t *this,
+/**
+ * Filter key exchange curves by curve user config
+ */
+static bool filter_curve_config(tls_named_group_t curve)
```
Besides filtering, ordering the DH groups might also be of interest. Maybe could be added as a new issue.

ryru replied:
You mean the order one would use in the config file for the key exchange TLS groups?

tobiasbrunner replied:
Yes, if one wanted to prefer _curve25519_ over _ecp256_. The order is currently determined by the plugin load order and their immutable internal algorithm registration order.

ryru replied:
New issue created: #23

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -1091,6 +1091,36 @@ static void filter_specific_config_suites(private_tls_crypto_t *this,
+
+			tok = lib->proposal->get_token(lib->proposal, token);
+			if (tok != NULL &&
+				tok->type == DIFFIE_HELLMAN_GROUP &&
```
```suggestion
			if (tok && tok->type == DIFFIE_HELLMAN_GROUP &&
```

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -1554,12 +1584,46 @@ CALLBACK(group_filter, bool,
+	VA_ARGS_VGET(args, out, curve);
+
+	while (orig->enumerate(orig, &group, &plugin))
+	{
```
That doesn't look correct. The underlying enumerator (which is the one returned by the inner `enumerator_create_filter()`) does not return a group and plugin but a group and a curve (i.e. it has basically the same signature as the hereby created enumerator). Then to filter you only have to call `filter_curve_config()`, no inner `for` loop or conversion of group to curve necessary.

tobiasbrunner commented on src/libtls/tls_crypto.h:
```diff
@@ -669,6 +669,14 @@ tls_crypto_t *tls_crypto_create(tls_t *tls, tls_cache_t *cache);
+ * @param curves		pointer to allocated curves array, to free(), or NULL
+ * @return				number of curves supported
+ */
+int tls_crypto_get_supported_curves(diffie_hellman_group_t **curves);
```
Since this now returns an array of internal DH group identifiers, the name of the function and variable might not be appropriate anymore.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -2322,6 +2386,25 @@ int tls_crypto_get_supported_suites(bool null, tls_version_t version,
+{
+	int count, i;
+
+	count = countof(curves);
```
As mentioned in another comment, you should enumerate over the DH groups supported by plugins and filter them so only those currently supported in TLS are returned.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -1216,7 +1252,7 @@ static status_t send_server_key_exchange(private_tls_server_t *this,
 	if (diffie_hellman_group_is_ec(group))
 	{
-		curve = ec_group_to_curve(this, group);
+		curve = tls_ec_group_to_curve(group);
```
You should probably move this change to the first commit in this branch (also the removal of the `ec_group_to_curve()` helper).

Using the new helper is not exactly the same, though. The previous code indirectly made sure that the DH group is actually supported. But that was probably redundant, because the group we get here initially will be the DH group from the selected cipher suite and suites with unsupported DH groups should have been filtered out earlier anyway.

tobiasbrunner commented on src/libtls/tests/suites/test_socket.c:
```diff
@@ -559,8 +651,26 @@ Suite *socket_suite_create()
 #define add_tls_curve_test(func) \
 	tcase_add_loop_test(tc, func, 0, tls_crypto_get_supported_curves(NULL));
 
+#define add_tls_versions_test(func, from, to) \
+	tcase_add_loop_test(tc, func, from, to);
+
```
No need for this macro, just use `tcase_add_loop_test()` directly.

tobiasbrunner commented on src/libtls/tests/suites/test_socket.c:
```diff
@@ -559,8 +651,26 @@ Suite *socket_suite_create()
 	tcase_add_loop_test(tc, func, 0, tls_crypto_get_supported_curves(NULL));
 
+#define add_tls_versions_test(func, from, to) \
+	tcase_add_loop_test(tc, func, from, to);
```
Like this the macro is not necessary, there is no benefit over calling `tcase_add_loop_test()` directly. However, you could perhaps change it to 
```suggestion
	tcase_add_loop_test(tc, func, from, to+1);
```
Then the versions you use below would actually be those tested.

tobiasbrunner commented on src/libtls/tests/suites/test_socket.c:
```diff
@@ -493,6 +493,98 @@ static void test_tls_curves(tls_version_t version, uint16_t port, bool cauth,
+{
+	echo_server_config_t *client, *server;
+
+	INIT(client,
+		.version = i,
+		.addr = "127.0.0.1",
+		.port = port,
+		.cauth = cauth,
+		.data = chunk_from_chars(0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08),
+	);
```
Maybe a constructor to create such a config would make sense to avoid having to duplicate all these lines.

tobiasbrunner commented on src/libtls/tests/suites/test_socket.c:
```diff
@@ -499,8 +556,16 @@ Suite *socket_suite_create()
 	tcase_add_loop_test(tc, func, 0, \
 						tls_crypto_get_supported_suites(TRUE, version, NULL));
 
+#define add_tls_curve_test(func) \
```
No need to define a macro as it's only called once.

tobiasbrunner commented on src/libtls/tests/suites/test_socket.c:
```diff
@@ -407,6 +407,21 @@ static void run_echo_client(echo_server_config_t *config)
+ * Configuration factory
+ */
+static void config_factory(tls_version_t version, uint16_t port, bool cauth,
+						   echo_server_config_t **config)
```
We generally use the regular constructor pattern for such functions (the verb probably more often at the beginning because it's a static helper), i.e. something like
```suggestion
static echo_server_config_t *create_config(tls_version_t version,
										   uint16_t port, bool cauth)
```

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -1554,12 +1583,41 @@ CALLBACK(group_filter, bool,
+
+	VA_ARGS_VGET(args, out, curve);
+
+	while (orig->enumerate(orig, &group, curve))
```
Why didn't you use the same approach for `group` and `curve`? In particular because the `group`-like approach is safer as according to the implementation of the existing filter, both arguments are optional. Sure, callers currently only set `group` to NULL but it's inconsistent like this and prone to problems in the future (the caller might also not expect the passed variables to change for unsuccessful enumerations).

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -1554,12 +1583,41 @@ CALLBACK(group_filter, bool,
+	return FALSE;
+}
+
+static enumerator_t* get_filtered_ec_enumerator()
```
This is not necessary, see comment for `tls_crypto_get_supported_groups()`.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -2322,6 +2380,36 @@ int tls_crypto_get_supported_suites(bool null, tls_version_t version,
+	tls_named_group_t curve;
+	int count = 0, i;
+
+	enumerator = get_filtered_ec_enumerator();
```
Actually, here only `group_filter` around the DH implementations is necessary (i.e. like the original `create_ec_enumerator()`. The whole point of splitting the filter was that this enumeration here will not be affected by the config (e.g. setting the global `libtls.curves` should not affect with which DH groups the unit tests are run).

ryru replied:
Ah, I see... I think I now understood it. Thank you for this elaboration!

tobiasbrunner commented on src/libstrongswan/crypto/proposal/proposal_keywords_static.txt:
```diff
@@ -1,5 +1,6 @@
 %{
 /*
+ * Copyright (C) 2020 Pascal Knecht
  * Copyright (C) 2009-2013 Andreas Steffen
  * HSR Hochschule fuer Technik Rapperswil
```
Seems this header hasn't been updated in a while:

```suggestion
 * Copyright (C) 2020 Pascal Knecht
 * Copyright (C) 2009-2019 Tobias Brunner
 * Copyright (C) 2009-2018 Martin Willi
 * Copyright (C) 2009-2017 Andreas Steffen
 * HSR Hochschule fuer Technik Rapperswil
```

tobiasbrunner commented on src/libtls/tests/suites/test_socket.c:
```diff
@@ -478,6 +481,72 @@ static void test_tls_curves(tls_version_t version, uint16_t port, bool cauth,
+	echo_server_config_t *client, *server;
+
+	config_factory(i, port, cauth, &client);
+	config_factory(i, port, cauth, &server);
```
```suggestion
	config_factory(version, port, cauth, &server);
```

tobiasbrunner commented on src/libtls/tests/suites/test_socket.c:
```diff
@@ -541,8 +610,26 @@ Suite *socket_suite_create()
+	add_tls_versions_test(test_tls_13_client, TLS_1_0, TLS_1_3);
+	suite_add_tcase(s, tc);
+
+	tc = tcase_create("TLS [1.0..1.2] client to TLS 1.2 server");
```
```suggestion
	tc = tcase_create("TLS [1.0..1.3] client to TLS 1.2 server");
```

tobiasbrunner commented on src/libtls/tests/suites/test_socket.c:
```diff
@@ -541,8 +610,26 @@ Suite *socket_suite_create()
+
+	tc = tcase_create("TLS 1.3 client to TLS [1.0..1.2] server");
+	tcase_add_checked_fixture(tc, setup_creds, teardown_creds);
+	add_tls_versions_test(test_tls_13_client, TLS_1_0, TLS_1_3);
```
```suggestion
	add_tls_versions_test(test_tls_13_client, TLS_1_0, TLS_1_2);
```
Or change the description so it says `[1.0..1.3] server` (might be kinda redundant, but why not test all combinations).
